NGX Datatable Defaults

Allcharles opened the pull request:
# Datatable Defaults

Created a directive for `<ngx-datatable>` components which loads the table with set defaults.

## Changes

- Created `src/app/directives/datatable/datatable.directive.ts
- To enable defaults the following must be added to a `<ngx-datatable>`

```typescript
<ngx-datatable
      #table
      appDatatable>
</ngx-datatable>
```

- To override defaults, you can do the following:

```typescript
<ngx-datatable
      #table
      appDatatable
      [defaults]=" { footerHeight: 150 } ">
</ngx-datatable>
```

- Updated Dependencies

## Closes

Closes #100

Allcharles committed:
Initial look at resolvers

Allcharles committed:
Initial look at resolvers

Allcharles committed:
Added resolvers key to pageInfo

Allcharles committed:
Added resolvers key to pageInfo

Allcharles committed:
Inject project model into project list

Allcharles committed:
Inject project model into project list

Allcharles committed:
Created error subroute

Allcharles committed:
Created error subroute

Allcharles committed:
Created site resolvers

Allcharles committed:
Created site resolvers

Allcharles committed:
Added api service resolvers

Allcharles committed:
Added api service resolvers

Allcharles committed:
Adjusting some pages to new system

Allcharles committed:
Adjusting some pages to new system

Allcharles committed:
Minor optimisations to boilerplate

Allcharles committed:
Minor optimisations to boilerplate

Allcharles committed:
Reducing size of combined resolvers

Allcharles committed:
Reducing size of combined resolvers

Allcharles committed:
Reduced size of inidividual resolvers

Allcharles committed:
Reduced size of inidividual resolvers

Allcharles committed:
Reduced unsubscribe boilerplate

Allcharles committed:
Abstract unsubscribe code to mixin

Allcharles committed:
Linted project

Allcharles committed:
Patch broken unit tests

Allcharles committed:
Defined WithUnsubscribe test headers

Allcharles committed:
Implemented unit tests

Allcharles committed:
Linted project

Allcharles committed:
Cleanup floating OnDestroy functions

Allcharles committed:
Created DeploymentEnvironmentService

- Created DeploymentEnvironmentService
- Added multiple fetch deployment environment attempts
- Added error message when deployment environment fails to load

Allcharles committed:
Cleanup typing

Allcharles committed:
Changed DeploymentEnvironment Service to appConfigService

Allcharles committed:
Fixed unit tests

Allcharles committed:
Fixed unit tests

Removed global services, this forces unit tests to actually utilise the helper functions which means mock services run instead of the dangerous live versions.

Allcharles committed:
Linted project

Allcharles committed:
Created DatatableDirective

Allcharles committed:
Initial datatable wrapper

Allcharles committed:
Updated permissions component

Allcharles committed:
Cleanup default datatable component

Allcharles committed:
Converted datatable wrapper to directive

Allcharles committed:
Created DatatableDirective unit tests

Allcharles committed:
Linted project

Allcharles committed:
Added default value overriding

Allcharles committed:
Merge pull request #125 from QutEcoacoustics/unsubscribe-mixin

Unsubscribe mixin

Allcharles committed:
Merge branch 'master' into app-deploy-environment

atruskie commented on src/app/directives/datatable/datatable.directive.ts:
```diff
@@ -0,0 +1,41 @@
+import { DatatableComponent } from "@swimlane/ngx-datatable";
+
+@Directive({
+  selector: "[appDatatable]"
```
suggested rename: `baw-datatable-defaults`

Allcharles replied:
Angular convention for directives is to prepend app so that it doesn't conflict with a library. My counter offer would be `appBawDatatableDefaults`

atruskie replied:
As per https://angular.io/guide/styleguide#directive-custom-prefix

Should be: `bawDatatableDefaults`

Allcharles replied:
I suspect that because angular ships with the following lint:
```
"directive-selector": [true, "attribute", "app", "camelCase"],
"component-selector": [true, "element", "app", "kebab-case"],
```
they want you to preface the directive and component with "app".
I've updated the rule to the following:

```
"directive-selector": [true, "attribute", ["app", "baw"], "camelCase"],
```

atruskie commented on src/app/directives/datatable/datatable.directive.ts:
```diff
@@ -0,0 +1,41 @@
+  selector: "[appDatatable]"
+})
+export class DatatableDirective implements OnInit, AfterContentInit {
+  @Input() datatable: DatatableComponent;
```
I believe you can use `@Host` here rather than `@Input` to get angular to inject whatever host component the directive is attached to

Allcharles committed:
List of changes:

- Removed analysis from build steps
- Added windows, mac, and linux VMs
- Moved build steps to linux
- Cached node_modules to reduce load times

atruskie commented on src/app/directives/datatable/datatable.directive.ts:
```diff
@@ -0,0 +1,41 @@
+
+  ngAfterContentInit() {
+    for (const key of Object.keys(this.datatableConfig)) {
+      this.datatable[key] = this.datatableConfig[key];
```
This is problematic. In your PR description, you said: "To override defaults". Defaults don't need to be overridden (do they?).

I'd suggest this is implemented like this:

1. get host component
1. `for property in this.datatableConfig`
    1. if the host's element has attribute of the same name, continue
    1. else, set the property on host

Allcharles replied:
Ngx-datatable comes with its own predefined defaults. Checking if a key exists before overwriting it would be hit and miss resulting in obscure issues.

atruskie replied:
I think you misunderstand.

Order of priority for inputs are (highest wins):

1. ngx defaults
2. our defaults
3. special-case override of our defaults <-- I don't think we need this
3. a value set by an attribute in the template

The rules are (highest wins):

1. if no default set by us, use theirs
2. if our default set, use that
3. if our default are overridden,  use those (<-- still don't think we need this)
4. if an attribute is set use it

With your current method, you could well see that an attribute is ignored because you always assign all of our defaults to the host component.

Rather, to respect rule 4, you should only assign one of our defaults to the host component IFF an attribute of the same name is not already set in the template.

Allcharles committed:
Merge branch 'master' into app-deploy-environment

atruskie reviewed (changes requested):
Looks pretty nice.

I'd suggest:
1. automatically getting host component (either via injection or `@Host`)
1. rename selector to `baw-datatable-defaults`
1. change the way overriding a default works

Allcharles committed:
Replaced datatable input with host

Allcharles committed:
Removed defaults from tableTemplate

Allcharles committed:
Merge pull request #126 from QutEcoacoustics/app-deploy-environment

App deploy environment

Allcharles committed:
Changed resolver names

Allcharles committed:
Changed resolver names

Allcharles committed:
Updated documentation

Allcharles committed:
Updated documentation

Allcharles committed:
Linted project

Allcharles committed:
Merge branch 'api-model-injection' of https://github.com/Allcharles/AngularTestbench into api-model-injection

Allcharles committed:
Updated profile pages

Allcharles committed:
Reduced duplication of resolvers list

Allcharles committed:
Created category parent tree

Allcharles committed:
Updated site pages

Allcharles committed:
Updated project pages

Allcharles committed:
Updated form submissions

Allcharles committed:
Cleanup project details unit tests

Allcharles committed:
Disabled failing unit tests

Allcharles committed:
Lint project

Allcharles committed:
Addressed review comments

Allcharles committed:
Merge pull request #116 from QutEcoacoustics/api-model-injection

API Model Injection

atruskie reviewed (changes requested):
This appears to have regressed from the last feedback.

I'd suggest deleting the commit and trying again

Allcharles committed:
Merge pull request #140 from QutEcoacoustics/simplify-CI

Simplify CI

Allcharles committed:
- Changed how errors are handled for resolvers. Resolver errors are now handled by the ResolverHandlerComponent. This simplifies the code in the error handler.
- Re-enable simple unit tests
- Re-enabled error-handler unit tests
- Re-enabled project list unit tests
- Re-enabled new project unit tests
- Re-enabled form unit tests
- Re-enabled new site unit tests
- Re-enabled login unit tests
- Re-enabled register unit tests
- Re-enabled unlock account unit tests
- Re-enabled reset password unit tests
- Re-enabled confirm account unit tests
- Re-enabled permissions-shield unit tests
- Re-enabled user badge unit tests
- Re-enabled my edit unit tests
- Re-enabled their profile unit tests
- Re-enabled their edit unit tests
- Re-enabled delete project unit tests
- Re-enabled project edit unit tests
- Re-enabled edit site unit tests
- Re-enabled delete site unit tests
- Re-enabled project permissions unit tests
- Re-enabled request  project unit tests
- Re-enabled site harvest unit tests
- Re-enabled project assign unit tests

Allcharles committed:
Removed ngAfterContentInit

Allcharles committed:
Merge pull request #142 from QutEcoacoustics/unit-tests

Re-enabled Unit Tests

Allcharles committed:
Merge branch 'datatable-wrapper-2' of https://github.com/Allcharles/AngularTestbench into datatable-wrapper-2

Allcharles committed:
Fixed toastr import

Allcharles committed:
Update dependencies

Allcharles committed:
Fixed audits
